### PR TITLE
feat(content-manager): add populate/fields/filters + depth guard to MCP read tools

### PR DESCRIPTION
**Pull request by sehrish30**
Improve the auto-generated MCP list_/get_ tools so an agent can assemble a page in one call instead of chasing bare { documentId } stubs.

- Add optional `fields`, `populate`, and `maxDepth` params to list_/get_ (collection + single types), mirroring the documents/entityService query shape. All optional; omitting them preserves current behavior.
- Opt-in relation inlining: relations named in `populate` are inlined as full entries, each sanitized against the RELATED type's own read permissions (per-target permissionChecker). Unreadable targets fall back to identity stubs. Inlining is one level deep; deeper relations stay stubbed.
- Response-size guard (`server.mcp.maxResponseBytes`, default 1 MB): oversized payloads return a truncated result with a clear notice instead of crashing.
- Extend filters to nested relation/component fields (one relation level deep), on top of the existing logical/field operators.
- Make list_'s shallow-omission explicit in the tool description.

### What does it do?

Adds optional fields, populate, and maxDepth params to the MCP list_/get_ read tools (mirroring the documents/entityService query shape). All optional and backward-compatible — omitting them keeps today's behavior.

populate inlines related entries in one call instead of returning { documentId } stubs. Inline depth follows the spec ({ author: { populate: ["avatar"] } } inlines author and author.avatar), and every level is sanitized against the related type's own read permissions; unreadable targets fall back to a stub.
fields / maxDepth control which scalar fields and how deep auto-populate goes.
Response-size guard (server.mcp.maxResponseBytes, default 1 MB) returns a truncated result with a notice instead of crashing on deep fetches.
Filters now support nested relation/component fields (e.g. { author: { name: { $contains: "Ada" } } }).
list_'s tool description now states relations return as stubs and sub-fields may be omitted at shallow depth.

### Why is it needed?

The MCP read tools were limited to point reads of a known entry:

No populate/fields control — list_ was fixed at depth 1, get_ at infinite depth, with nothing in between.
Deep get_ fetches on nested pages produced huge payloads with no cap, which could crash the MCP transport.
Relations always returned as bare { documentId } stubs, forcing 3–4 manual chase calls to read one page.
list_ silently omitted nested sub-fields (absent, not null), so emptiness couldn't be judged from list output.
list_ filters only supported top-level scalars.

### How to test it?

Automated (this PR adds/extends tests — 7 suites, 322 unit tests):

yarn --cwd packages/core/content-manager test:unit src/mcp
yarn --cwd packages/core/content-manager test:ts:back
yarn eslint packages/core/content-manager/server/src/mcp --ext .ts
Manual (getstarted sandbox with MCP enabled):

cd examples/getstarted && yarn develop, enable the MCP server, connect an MCP client.
get_<ct>({ documentId, populate: { author: { populate: ["avatar"] } } }) → author and author.avatar are inlined (RBAC-sanitized); unnamed relations stay { documentId } stubs.
list_<ct>({ fields: ["title"], filters: { <relation>: { <field>: { $contains: "…" } } } }) → nested filter applies; only requested fields return.
Set server.mcp.maxResponseBytes low and get_ a deeply-populated document → truncated result with a notice instead of a crash.
Call any read tool with no new params → identical to previous behavior.

### Related issue(s)/PR(s)
Let us know if this is related to any issue/pull request

### CODE REVIEW

File: packages/core/content-manager/server/src/mcp/schemas/filters-schema.ts [hunk 1]
```diff
@@ -56,39 +56,170 @@ export const FILTER_OPERATORS = [
...
+          if (nested !== undefined) nestedShapes[key] = nested.optional();
+        }
+      } else if (
+        attr.type === 'relation' &&
+        (attr as { target?: string }).target !== undefined &&
+        (attr as { target?: string }).target !== 'admin::user'
+      ) {
+        const target = getModel((attr as { target: string }).target);
+        if (target?.attributes !== undefined) {
+          const nested = buildNestedTargetFilter(target.attributes, getModel, new Set(), false);
```
Comment (nclsndr) [inline packages/core/content-manager/server/src/mcp/schemas/filters-schema.ts]: **Relation filters expose unreadable target fields (high)**

`buildFiltersSchema` expands target fields, but the handler sanitizes only with the source checker. A real permission-manager probe retained `author.secret.$eq` when source read was allowed and target read denied, exposing facts through result membership/count.

Please remove nested relation filtering from this PR while keeping component filtering. The existing shaping API fixture already has `mcp-shape-source.oneTarget` and target `secret` values. Create a token with source read only, then call `tools/call` with `{ name: 'list_mcp-shape-source', arguments: { filters: { oneTarget: { secret: { $eq: 'SECRET-target-1' } } } } }`. The current schema accepts this and the filtered result reveals the matching target. After the scope cut, assert `result.isError === true`. A focused follow-up must prove denied target-type and target-field access cannot affect membership while an explicitly permitted target field still works.

---
_AI assisted_

File: packages/core/content-manager/server/src/mcp/handlers/collection-handlers.ts [hunk 2]
```diff
@@ -90,16 +113,28 @@ export const createCollectionListHandler =
...
 
     const permissionQuery = await permissionChecker.sanitizedQuery.read(query);
 
-    const populate = await getService('populate-builder')(uid)
-      .populateFromQuery(permissionQuery)
-      .populateDeep(1)
+    // When `populate` is explicit, honor it exactly (populateFromQuery). Otherwise keep the
+    // legacy auto-populate to depth 1 (overridable via `maxDepth`). Inlining is opt-in and
+    // driven off the sanitized populate so RBAC-denied relations are never inlined.
+    const populateBuilder = getService('populate-builder')(uid).populateFromQuery(permissionQuery);
```
Comment (nclsndr) [inline packages/core/content-manager/server/src/mcp/handlers/collection-handlers.ts]: **Explicit populate is dropped before Document Service (high)**

`populateFromQuery(permissionQuery)` calls `getQueryPopulate`, which traverses filters rather than preserving explicit `populate`. Runtime probes turned both `['author']` and `'*'` into `{}`, and collection reads reached Document Service with `populate: {}`. Single get additionally replaces its computed shape with unconditional deep population.

Please forward the sanitized explicit shape, compose only the localization override, and use that same shape for inlining. The shaping suite already has `get_mcp-shape-source`, `oneTarget`, and target `name`/`secret`. Call `tools/call` with `{ name: 'get_mcp-shape-source', arguments: { documentId, populate: ['oneTarget'] } }` and assert `result.structuredContent.data.oneTarget.name === 'Target 1'`; this currently fails. The same fixture supports `'*'` and `{ oneTarget: { fields: ['name'] } }`. Also assert final collection Document Service parameters preserve each form. Single-type coverage needs a focused handler test or a new single-type API fixture because this suite has none.

---
_AI assisted_

File: packages/core/content-manager/server/src/mcp/schemas/query-schema.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    }
+    if (typeof node === 'object') {
+      for (const [key, value] of Object.entries(node as Record<string, unknown>)) {
+        if (value === false || value === undefined || value === null) {
+          // eslint-disable-next-line no-continue
+          continue;
+        }
+        const path = joinPath(prefix, key);
+        exact.add(path);
+        if (typeof value === 'object' && !Array.isArray(value) && 'populate' in value) {
```
Comment (nclsndr) [inline packages/core/content-manager/server/src/mcp/schemas/query-schema.ts]: **Fragment populate paths are never collected (medium)**

Canonical dynamic-zone and morph shapes place nested population under `on`, but this walker only follows a direct `populate` key. For example, `blocks.on['shared.hero'].populate.author` must match runtime path `blocks.author`, without adding the component UID. Current probes leave that author as an identity stub.

Please traverse `on` fragments and first pin the matcher in `query-schema.test.ts`: `{ blocks: { on: { 'shared.hero': { populate: ['author'] } } } }` should make `shouldInline('blocks.author')` true; it currently returns false. End-to-end coverage requires new fixtures: the shaping API source currently has no dynamic zone/component, and its morph targets have no nested relation. Add a dynamic-zone component containing an author relation and invoke the generated get tool through `tools/call`, asserting the readable author field is present and restricted fields remain absent. Morph coverage also needs registered input-schema plumbing because the current `morphTargets` attribute is omitted from the Content Manager model DTO.

---
_AI assisted_

File: packages/core/content-manager/server/src/mcp/handlers/collection-handlers.ts [hunk 4]
```diff
@@ -163,14 +211,28 @@ export const createCollectionGetHandler =
       throw new errors.ForbiddenError();
     }
 
-    const permissionQuery = await permissionChecker.sanitizedQuery.read({ locale, status });
+    const permissionQuery = await permissionChecker.sanitizedQuery.read({
+      locale,
+      status,
+      ...(fields !== undefined && { fields }),
```
Comment (nclsndr) [inline packages/core/content-manager/server/src/mcp/handlers/collection-handlers.ts:188]: **Collection get sanitizes fields but never applies them (medium)**

`fields` enters `permissionChecker.sanitizedQuery.read`, but the later `documentManager.findOne` options contain only `populate`, locale, and status. Output sanitization enforces RBAC, not the caller's projection, so it cannot compensate; a runtime request for `fields: ['title']` returned another readable scalar.

Please pass the sanitized projection to `findOne`. The existing RBAC fixture already has `mcp-rbac-doc.title` and `secret`; no schema changes are needed. Seed both values, grant the token read permission for both fields, and call `tools/call` with `{ name: 'get_mcp-rbac-doc', arguments: { documentId, fields: ['title'] } }`. Assert the call succeeds, `result.structuredContent.data.title === 'visible'`, and `data` has no `secret` property. The last assertion currently fails because `secret` is readable but was not requested.

---
_AI assisted_

File: packages/core/content-manager/server/src/mcp/schemas/query-schema.ts [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+ *
+ * Mirrors Strapi's entityService/documents `fields` query parameter. Constraining the
+ * enum to permitted scalar keys keeps RBAC field filtering intact. Returns `z.never()`
+ * when the model exposes no readable scalar fields.
+ */
+export const buildFieldsSchema = (
+  attributes: Struct.SchemaAttributes,
+  permittedFields?: Set<string> | null
+): z.ZodTypeAny => {
+  const scalarKeys = getScalarAttributeKeys(attributes, permittedFields);
```
Comment (nclsndr) [inline packages/core/content-manager/server/src/mcp/schemas/query-schema.ts]: **JSON and Blocks are valid projected scalars (medium)**

`buildFieldsSchema` reuses the sort/filter scalar allowlist, which intentionally excludes JSON and Blocks. Both are valid Document Service projections, so `fields: ['payload']` and `fields: ['body']` currently fail MCP input validation before reaching the handler.

Please use a read-specific projectable-field helper without widening sort eligibility. Add `payload: { type: 'json' }` and `body: { type: 'blocks' }` to the existing `mcp-rbac-doc` API fixture; these attributes do not exist there today. Seed a JSON value and a valid Blocks paragraph, grant read permission for both, then call `tools/call` on `get_mcp-rbac-doc` with `arguments: { documentId, fields: ['payload'] }` and separately `fields: ['body']`. Assert `result.isError !== true` and the requested value exists under `result.structuredContent.data`; the success assertion currently fails during input validation. Add matching `buildFieldsSchema` unit assertions for both types.

---
_AI assisted_

File: packages/core/content-manager/server/src/mcp/schemas/query-schema.ts [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+ */
+export const buildMaxDepthSchema = (): z.ZodTypeAny =>
+  z
+    .number()
+    .int()
+    .min(0)
+    .max(10)
+    .optional()
+    .describe(
+      'Max relation-population depth for auto-populate (when "populate" is omitted). ' +
```
Comment (nclsndr) [inline packages/core/content-manager/server/src/mcp/schemas/query-schema.ts]: **maxDepth does not provide the advertised relation depth (medium)**

The schema advertises relation-population depth, but `getDeepPopulate` increments depth only through components and dynamic zones. Ordinary relations are emitted as `true`, so real builder probes at depths 1 and 3 produce the same populate shape.

Please remove `maxDepth` from this PR, including handler arguments, generated schemas, descriptions, and mocked wiring tests. After the scope cut, assert generated collection list/get and single-get input schemas no longer expose or accept it. A focused follow-up needs new depth fixtures: existing populate tests have isolated relation models, and the shaping API target has no nested relation. Add a source → target → nested-target chain, compare the final populate sent to Document Service for depths 1 and 3, and assert depth 3 contains the nested ordinary relation while depth 1 does not. That distinction fails today because both stop at `{ relation: true }`.

---
_AI assisted_

File: packages/core/content-manager/server/src/mcp/utils.ts [hunk 7]
```diff
@@ -53,10 +57,121 @@ export const sanitizeFormatShape = async (
...
+export const enforceResponseBudget = (
+  structuredContent: Record<string, unknown>,
+  maxBytes: number,
+  buildTruncated: (notice: string) => Record<string, unknown>
+): Record<string, unknown> => {
+  if (maxBytes <= 0) {
+    return structuredContent;
+  }
+
+  const size = Buffer.byteLength(JSON.stringify(structuredContent), 'utf8');
```
Comment (nclsndr) [inline packages/core/content-manager/server/src/mcp/utils.ts]: **This budget does not cap the transported response (medium)**

`enforceResponseBudget` measures only `structuredContent`; `ok()` then serializes the same object again as text, and JSON-RPC/SSE adds framing. A 999,920-byte guarded value produced 1,999,906 handler bytes and 1,999,963 SSE bytes. The fallback is not rechecked, and this post-fetch check cannot limit query or memory amplification.

Please remove the guard from this PR. A focused follow-up can reuse the shaping suite's `/mcp` transport and `mcp-shape-source.name`; no new content type is needed, but it needs a configured large-payload case. Invoke raw `mcpRpc`—before `parseMcpResponse`—with method `tools/call` and params `{ name: 'get_mcp-shape-source', arguments: { documentId, fields: ['name'] } }`, then assert `Buffer.byteLength(res.text, 'utf8')` stays within the configured transport cap. That assertion currently exceeds the cap. Also define and test behavior when even the fallback envelope is too large, and expose the config publicly.

---
_AI assisted_

**Review by nclsndr — CHANGES_REQUESTED** [review]
Thank you so much @sehrish30 for enhancing the MCP capabilities!

Here a preliminary review before QA. 

> [!NOTE]
>  If you ever lack bandwidth to carry on on this feature, please let us know, strapi team would be happy to take over.

---

The main request is scope reduction: please split this work along its five independently shippable feature lines—filters, `populate` plus RBAC-safe relation inlining, `fields`, `maxDepth`, and the response-size guard. For this PR, I recommend keeping `fields`, explicit `populate`, RBAC-safe inlining, and component filtering; defer nested relation filters, `maxDepth`, and the response-size guard to focused follow-ups. The issues on those three deferred features are scope-cut requests, not requests to patch them in place here.

The retained features still need fixes before merging:
- Explicit `populate` is not preserved to Document Service, and canonical dynamic-zone/morph `on` fragments do not drive nested inlining.
- Collection get sanitizes `fields` but does not pass them to `findOne`; JSON and Blocks projections are rejected by the sort-oriented scalar allowlist.

This is a narrow file footprint, but it bundles five separate contracts and the confirmed defects track those feature boundaries almost one-for-one. The current unit coverage mocks the permission, populate-builder, Document Service, and transport boundaries, so please add real MCP/API regressions for target permissions, exact Document Service query propagation, and shaping. If the response guard returns later, its config must also be public, typed, and documented.

---
_AI assisted_

File: packages/core/content-manager/server/src/mcp/schemas/filters-schema.ts (hunk 1 above)
Comment (sehrish30) [reply packages/core/content-manager/server/src/mcp/schemas/filters-schema.ts]: Fixed in cd1eb3922c. Removed relation-target expansion from buildNestedTargetFilter/buildFiltersSchema entirely — component filtering is preserved, relation targets are never expanded into nested filters anymore.

Added the exact regression you described in tests/api/core/mcp/mcp-content-manager-shaping.test.api.ts ("nested relation filters are rejected (scope cut)"): filters: { oneTarget: { secret: { $eq: ... } } } against mcp-shape-source/mcp-shape-target now asserts result.isError === true, with a control test confirming top-level scalar filtering still works.

File: packages/core/content-manager/server/src/mcp/handlers/collection-handlers.ts (hunk 2 above)
Comment (sehrish30) [reply packages/core/content-manager/server/src/mcp/handlers/collection-handlers.ts]: Fixed in cd1eb3922c. populateFromQuery still only walks filters/sort (unchanged core behavior), so instead of routing explicit populate through it, both collection list/get and single-type get now forward the sanitized populate shape as-is via a new composePopulate helper (mcp/utils.ts), composed only with the localizations override. The single get handler no longer force-overrides with unconditional deep populate when populate is explicit.

Added tests/api/core/mcp/mcp-content-manager-shaping.test.api.ts ("explicit populate propagation"): array form, '*', and nested { fields } form against get_mcp-shape-source/list_mcp-shape-source, asserting oneTarget.name === 'Target 1' is actually inlined (previously failed exactly as you described). Also added collection Document Service parameter assertions at the unit level in derive-content-type-mcp-tools.test.ts.

One thing this surfaced once tests ran for real: inlined relations were leaking locale: null for non-localized targets and failing output-schema validation — fixed in 32618c1d9d.

File: packages/core/content-manager/server/src/mcp/schemas/query-schema.ts (hunk 3 above)
Comment (sehrish30) [reply packages/core/content-manager/server/src/mcp/schemas/query-schema.ts]: Fixed in cd1eb3922c. buildInlinePathMatcher now traverses on fragments and collects them against the parent path (not the component UID) — { blocks: { on: { 'shared.hero': { populate: ['author'] } } } } now makes shouldInline('blocks.author') return true.

Added the fixture you flagged was missing: mcp-shape-source now has a blocks dynamic zone with a new mcp-shape-blocks.hero component containing an author relation (tests/api/core/mcp/mcp-content-manager-shaping.test.api.ts). New test calls get_mcp-shape-source through tools/call with the on-wrapped populate and asserts the readable author.name is present while secret stays absent; a control test confirms populate: ['blocks'] alone never reaches the nested relation at all (no on = never fetched, not just un-inlined). Also added unit coverage in query-schema.test.ts.

(Morph coverage / registered input-schema plumbing for morphTargets in the CM model DTO is still open — that one I haven't touched.)

File: packages/core/content-manager/server/src/mcp/handlers/collection-handlers.ts (hunk 4 above)
Comment (sehrish30) [reply packages/core/content-manager/server/src/mcp/handlers/collection-handlers.ts:188]: Fixed in cd1eb3922c. The sanitized fields value is now spread into documentManager.findOne's options.

Added the exact scenario from your repro in tests/api/core/mcp/mcp-content-manager-rbac.test.api.ts: seeded mcp-rbac-doc.title/secret with read permission on both, called get_mcp-rbac-doc with fields: ['title'], asserted data.title === 'visible' and data has no secret property (isolates the projection bug from RBAC, since both fields are readable).

File: packages/core/content-manager/server/src/mcp/schemas/query-schema.ts (hunk 5 above)
Comment (sehrish30) [reply packages/core/content-manager/server/src/mcp/schemas/query-schema.ts]: Fixed in cd1eb3922c. Added a read-specific getProjectableAttributeKeys/PROJECTABLE_ATTRIBUTE_TYPES in sort-schema.ts (superset of the sort/filter allowlist, adding json/blocks) and switched buildFieldsSchema to use it — sort/filter eligibility is untouched.

Added payload: json and body: blocks to the mcp-rbac-doc fixture as you suggested, plus a test seeding both and calling get_mcp-rbac-doc with fields: ['payload'] then fields: ['body'] — both succeed and return the seeded values. Unit assertions added in query-schema.test.ts too.

File: packages/core/content-manager/server/src/mcp/schemas/query-schema.ts (hunk 6 above)
Comment (sehrish30) [reply packages/core/content-manager/server/src/mcp/schemas/query-schema.ts]: Removed per your scope-cut request in cd1eb3922c — maxDepth is gone entirely: schema, handler args (collection list/get, single get), wiring in derive-content-type-mcp-tools.ts, and tool descriptions. Will file a follow-up issue to reintroduce it properly once getDeepPopulate actually increments depth through plain relations, not just components/dynamic zones, with the depth-1-vs-depth-3 fixture comparison you described.

File: packages/core/content-manager/server/src/mcp/utils.ts (hunk 7 above)
Comment (sehrish30) [reply packages/core/content-manager/server/src/mcp/utils.ts]: Removed per your scope-cut request in cd1eb3922c — enforceResponseBudget/getMaxResponseBytes/MCP_DEFAULT_MAX_RESPONSE_BYTES are gone, along with the truncated/notice output-schema fields they populated. Will file a follow-up issue to reintroduce it measuring the actual transported bytes (post-serialization, post-framing) rather than just structuredContent, with public/typed config as you asked.

**Comment by sehrish30** [comment]
> Thank you so much @sehrish30 for enhancing the MCP capabilities!
> 
> Here a preliminary review before QA.
> 
> Note
> 
> If you ever lack bandwidth to carry on on this feature, please let us know, strapi team would be happy to take over.
> 
> The main request is scope reduction: please split this work along its five independently shippable feature lines—filters, `populate` plus RBAC-safe relation inlining, `fields`, `maxDepth`, and the response-size guard. For this PR, I recommend keeping `fields`, explicit `populate`, RBAC-safe inlining, and component filtering; defer nested relation filters, `maxDepth`, and the response-size guard to focused follow-ups. The issues on those three deferred features are scope-cut requests, not requests to patch them in place here.
> 
> The retained features still need fixes before merging:
> 
> * Explicit `populate` is not preserved to Document Service, and canonical dynamic-zone/morph `on` fragments do not drive nested inlining.
> * Collection get sanitizes `fields` but does not pass them to `findOne`; JSON and Blocks projections are rejected by the sort-oriented scalar allowlist.
> 
> This is a narrow file footprint, but it bundles five separate contracts and the confirmed defects track those feature boundaries almost one-for-one. The current unit coverage mocks the permission, populate-builder, Document Service, and transport boundaries, so please add real MCP/API regressions for target permissions, exact Document Service query propagation, and shaping. If the response guard returns later, its config must also be public, typed, and documented.
> 
> _AI assisted_

Thanks for the thorough review! Addressed everything as scoped:

Scope cut: removed maxDepth, the response-size guard, and nested relation filtering from this PR (kept component filtering). Will file follow-up issues for the three deferred features as requested.

Explicit populate: now forwarded to the Document Service unchanged via a new composePopulate helper (packages/core/content-manager/server/src/mcp/utils.ts), composed only with the localizations override.
fields on collection get: now passed into documentManager.findOne.

Real regressions added: tests/api/core/mcp/mcp-content-manager-rbac.test.api.ts and tests/api/core/mcp/mcp-content-manager-shaping.test.api.ts now exercise real permission-manager probes, exact Document Service query propagation, and shaping — including your exact repro fixtures (mcp-shape-source.oneTarget, mcp-rbac-doc.title/secret) plus a new dynamic-zone/component fixture and single-type coverage.

One more bug surfaced while writing those tests for real: inlined relations were leaking a raw locale: null for non-localized targets, failing output-schema validation. Fixed in the same pass, with unit test coverage.

Commits: cd1eb3922c (fixes), 9c98020c0b (tests), 32618c1d9d (locale-leak fix found by the new tests). All 32 API tests + 317 unit tests pass; lint/tsc clean.

**Comment by sehrish30** [comment]
Hi @nclsndr !Just a friendly follow-up on this PR. I’ve addressed the review comments and replied a few weeks ago. Whenever you get a chance, could you please take another look? Thanks!

**Comment by nclsndr** [comment]
> Hi @nclsndr !Just a friendly follow-up on this PR. I’ve addressed the review comments and replied a few weeks ago. Whenever you get a chance, could you please take another look? Thanks!

Sorry for the latency @sehrish30. We've had new bug reports on the MCP that I'll have to prioritize before getting back to this PR. I'll make sure it get's cited on GitHub so you'll have the tracking

**Comment by sehrish30** [comment]
> > Hi @nclsndr !Just a friendly follow-up on this PR. I’ve addressed the review comments and replied a few weeks ago. Whenever you get a chance, could you please take another look? Thanks!
> 
> Sorry for the latency @sehrish30. We've had new bug reports on the MCP that I'll have to prioritize before getting back to this PR. I'll make sure it get's cited on GitHub so you'll have the tracking



> > Hi @nclsndr !Just a friendly follow-up on this PR. I’ve addressed the review comments and replied a few weeks ago. Whenever you get a chance, could you please take another look? Thanks!
> 
> Sorry for the latency @sehrish30. We've had new bug reports on the MCP that I'll have to prioritize before getting back to this PR. I'll make sure it get's cited on GitHub so you'll have the tracking

Thanks for the update! No worries, completely understand. I’ll keep an eye out for the tracking issue. Happy to continue iterating on the PR when there’s bandwidth to revisit it.

**Comment by nclsndr** [comment]
For the record, the issue that stops current MCP merges.
- #27395
First requirement was Zod >= 4.2
- #27428
Now working on upgrading mcp sdk to v2.x

**Comment by sehrish30** [comment]
> For the record, the issue that stops current MCP merges.
> 
> * [MCP `tools/list` advertises draft-07 JSON Schema, making all tools uncallable from strict 2020-12 clients #27395](https://github.com/strapi/strapi/issues/27395)
>   First requirement was Zod >= 4.2
> * [chore(deps): migrate first-party zod to 4.4.3 #27428](https://github.com/strapi/strapi/pull/27428)
>   Now working on upgrading mcp sdk to v2.x

merged develop in and re-ran the full suite — 321/321 unit tests and 32/32 MCP API tests passing. One thing the merge surfaced: the shaping test suite's setup (component + content types via CTB API + full boot + locales) started exceeding Jest's default 60s hook timeout post-merge — bumped it to 120s, unrelated to the actual read-tool logic.

Still watching for the MCP SDK v2.x upgrade before doing another pass.

**Comment by nclsndr** [comment]
The schema-dialect regression that blocked this work has a fix PR: https://github.com/strapi/strapi/pull/27505

Please re-review after that merge.

---
_AI assisted_

**Comment by sehrish30** [comment]
> The schema-dialect regression that blocked this work has a fix PR: #27505
> 
> Please re-review after that merge.
> 
> _AI assisted_

Thanks for the pointer — I'll keep an eye on #27505 and merge develop in again once it lands, then re-run the full suite and ping for re-review.

**Comment by nclsndr** [comment]
Thank you for your patience on this one, @sehrish30.

#27505 is now merged. The schema-dialect problem that blocked all MCP merges is fixed, so this PR is no longer held by it.

The branch now has conflicts with `develop`.

Do you want to resolve the conflicts and continue with this PR, or do you prefer that the Strapi team takes it over? Both are fine — tell us which one you prefer.

---
_AI assisted_

**Comment by sehrish30** [comment]
> Thank you for your patience on this one, @sehrish30.
> 
> #27505 is now merged. The schema-dialect problem that blocked all MCP merges is fixed, so this PR is no longer held by it.
> 
> The branch now has conflicts with `develop`.
> 
> Do you want to resolve the conflicts and continue with this PR, or do you prefer that the Strapi team takes it over? Both are fine — tell us which one you prefer.
> 
> _AI assisted_

Resolved the conflict with develop (one collision in mcp-content-manager-rbac.test.api.ts — the schema-dialect fix's new publishedOn test field vs. my payload/body/single-type additions, now combined). Branch is fully caught up with develop. Re-ran the full suite now and pushed once confirmed green.
